### ISSUE TITLE
[FastCGI] Chunked HTTP Answer is Incomplete

### ISSUE THREAD

**Issue by kamulos** [haproxy/haproxy#886]
## Detailed description of the problem

I have a FastCGI server, that returns multiple KB answers regularly. Haproxy returns this via HTTP chunked encoding. There are sporadic errors, where the HTTP answer of haproxy is incomplete. At least the end chunk is missing from the HTTP answer to which libcurl responds with the following error:

```
transfer closed with outstanding read data remaining
```
This sound very similar to issue #231 

It affects every version from 2.1.0 until 2.3 git

## Steps to reproduce the behavior

The issue definitely occurs more or less often depending on the size of the answer. For example if I choose a buffer size of 65536, I can not reproduce it at all.

### FastCGI App

You need the libfcgi and spawn-fcgi to run this

#### fcgiserver.c
``` C
#include <fcgi_stdio.h>

char buffer[31592];

int main(void)
{
  while (FCGI_Accept() >= 0)   {
    printf("\r\n");
    fwrite(buffer, sizeof(buffer), 1, stdout);
    fflush(stdout);
  }
}
```
#### compile
```
cc fcgiserver.c -o fcgiserver -lfcgi
```
#### run
```
spawn-fcgi -n -p 3452 ./fcgiserver
```

### HAProxy Config
```
defaults
	timeout connect 5000
	timeout client  50000
	timeout server  50000

frontend reverser-http
	bind *:8080
	mode http
	use_backend errortest

backend errortest
	mode http
	use-fcgi-app fcgi-errortest
	server myserver 127.0.0.1:3452 proto fcgi

fcgi-app fcgi-errortest
	no option keep-conn
	docroot /
```

### Client

This needs the libcurl and will just request the FastCGI answer in a loop

#### client.c

``` C
#include <curl/curl.h>

static size_t write_data(void *buffer, size_t size, size_t nmemb, void *userp)
{
  return size * nmemb;
}

int main(void)
{
  int ret = 0;

  CURL *connection = NULL;
  CURLcode res = CURLE_OK;

  char errorstr[CURL_ERROR_SIZE];

  if (curl_global_init(CURL_GLOBAL_ALL) != 0) {
    ret = 1;
    goto EXIT_POINT;
  }

  connection = curl_easy_init();

  if (connection == NULL) {
    ret = 2;
    goto EXIT_POINT;
  }

  curl_easy_setopt(connection, CURLOPT_URL, "http://127.0.0.1:8080/");
  curl_easy_setopt(connection, CURLOPT_WRITEFUNCTION, write_data);
  curl_easy_setopt(connection, CURLOPT_FAILONERROR, 1);
  curl_easy_setopt(connection, CURLOPT_ERRORBUFFER, errorstr);
  curl_easy_setopt(connection, CURLOPT_FORBID_REUSE, 1L);

  for (size_t i=0;;i++) {
    errorstr[0] = '\0';

    res = curl_easy_perform(connection);

    if (res != CURLE_OK) {
      printf("CURL ERROR: %s\n", errorstr);
      ret = 3;
      goto EXIT_POINT;
    }

    printf("%zu\n", i);
  }

EXIT_POINT:

  if (connection != NULL) {
    curl_easy_cleanup(connection);
  }

  curl_global_cleanup();

  return ret;
}
```

#### compile
```
cc client.c -o client -lcurl
```

#### run
```
./client
```

**Comment by capflam** [comment]
I'm able to reproduce the bug. The end of stream is reported too early, while there is still unparsed data. This happens because a shutdown happens at the connection level. Here is a patch that should fix your bug. It may be applied on any version from the 2.3-dev to the 2.1 : 

[0001-BUG-MEDIUM-mux-fcgi-Report-EOS-when-no-more-data-are.patch.txt](https://github.com/haproxy/haproxy/files/5339122/0001-BUG-MEDIUM-mux-fcgi-Report-EOS-when-no-more-data-are.patch.txt)

I want to review it more carefully before pushing it to be sure it is the right way to fix the bug and it does not introduced other subtle bugs (like a blocked stream for instance). But it really fixes the bug I observed and most probably yours too.

Thanks !

**Comment by kamulos** [comment]
Thanks! With this patch I can not reproduce the bug neither with the minimal example above nor with my original program.

**Comment by capflam** [comment]
Good ! Thanks for the feedback.

**Comment by capflam** [comment]
It should be fixed now in 2.3-DEV. I'll backport the fix soon. Thanks !

**Comment by capflam** [comment]
done !
